From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Thu, 24 Apr 2025 23:11:13 +0800
Subject: [PATCH] Luminol Fix off tickregion sync teleport

Co-authored by: MrHua269 <mrhua269@gmail.com>
As part of: Luminol (https://github.com/LuminolMC/Luminol/blob/ver/1.21.8/luminol-server/minecraft-patches/features/0061-Fix-off-tickregion-sync-teleport.patch)
Licensed under: MIT (https://github.com/LuminolMC/Luminol/blob/ver/1.21.8/licenses/MIT.md)

diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index e63d991747f4e769125e11ceefae0d37b5c55a7d..269c1aa1c137f9e8d85fd5ce7c70172aed81dc20 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4312,6 +4312,21 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         this.resetStoredPositions();
     }
 
+    // Luminol start - Fix sync teleport issue
+    private boolean checkNearbyTickRegions(int destX, int destZ) {
+        // Dumb end gateway search the chunks in radius of 5 chunks, so we need keep 6(5+1) by default check radius
+        int sizeBx = Math.min(6, (int) (this.bb.maxX - this.bb.minX) + this.level.getCraftServer().getSimulationDistance());
+        int sizeBz = Math.min(6, (int) (this.bb.maxZ - this.bb.minZ) + this.level.getCraftServer().getSimulationDistance());
+
+        // check tick thread around these area
+        return ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.level,
+                (destX >> 4) - sizeBx,
+                (destZ >> 4) - sizeBz,
+                (destX >> 4) + sizeBx,
+                (destZ >> 4) + sizeBz);
+    }
+    // Luminol end
+
     protected final void transform(TeleportTransition telpeort) {
         PositionMoveRotation move = PositionMoveRotation.calculateAbsolute(
             PositionMoveRotation.of(this), PositionMoveRotation.of(telpeort), telpeort.relatives()
@@ -4409,7 +4424,8 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         // check for same region
         if (destination == this.level()) {
             Vec3 currPos = this.position();
-            if (
+            // Luminol - Prevent entity sync teleported to the edge of tickregion
+            if (this.checkNearbyTickRegions((int) pos.x, (int) pos.z) && // Luminol - Fix sync teleport issue
                 destination.regioniser.getRegionAtUnsynchronised(
                     ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkX(currPos), ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkZ(currPos)
                 ) == destination.regioniser.getRegionAtUnsynchronised(
